packages/preview2-shim/lib/io/worker-io.js

```diff
@@ -0,0 +1,205 @@
+import { fileURLToPath } from 'node:url';
+import { createSyncFn } from '../synckit/index.js';
+import * as calls from './calls.js';
+import * as _streamTypes from './stream-types.js';
+
+const DEBUG = false;
+
+const workerPath = fileURLToPath(new URL('./worker-thread.js', import.meta.url));
+
+/**
+ * @type {(call: number, id: number | null, payload: any) -> any}
+ */
+let ioCall = createSyncFn(workerPath);
+if (DEBUG) {
+  const _ioCall = ioCall;
+  ioCall = function ioCall(num, id, payload) {
+    console.trace((num & calls.CALL_MASK) >> calls.CALL_SHIFT, num & calls.CALL_TYPE_MASK, id, payload);
+    return _ioCall(num, id, payload);
+  }
+}
+
+export { _streamTypes }
+
+const symbolDispose = Symbol.dispose || Symbol.for('dispose');
+
+class ComponentError extends Error {
+  constructor (name, msg, payload) {
+    super(msg);
+    this.name = name;
+    this.payload = payload;
+  }
+}
+
+const _Error = Error;
+const IoError = class Error extends _Error {
+  toDebugString () {
+    return this.message;
+  }
+}
+
+function streamIoErrorCall(call, id, payload) {
+  try {
+    return ioCall(call, id, payload);
+  } catch (e) {
+    // any invalid error is a trap
+    if (e.tag !== 'stream-error') {
+      console.error(e);
+      process.exit(1);
+    }
+    if (e.val.tag === 'last-operation-failed') {
+      const msg = e.val.val.message;
+      e.val.val = Object.assign(new IoError(), e.val);
+      throw new ComponentError("StreamError", msg, e);
+    }
+    throw new ComponentError("StreamError", "stream closed", e);
```

Comment:
It would be interesting to provide a single implementation for error handling that we could use across all code. wdyt?

Reply:
I would be open to any refactoring here, still kind of figuring it out myself!

Reply:
This is not urgent and can be done later when we finish laying the foundation for the other implementations.